File: Algorithms/Greedy/Jobs.java

import java.util.*;

public class Jobs {

    static class Job implements Comparable<Job> {

        String name;
        int deadline, profit;

        public Job(String n, int d, int p) {
            name = n;
            deadline = d;
            profit = p;
        }

        @Override
        public int compareTo(Job b) {
            return b.profit - profit;
        }

        @Override
        public String toString() {
            return name;
        }

    }

    public static void printBestOrder(List<Job> jobs) {

        Collections.sort(jobs);

        int n = jobs.size();

        int[] results = new int[n];
        boolean[] times = new boolean[n];

        for (int i = 0; i < n; i++) {

            for (int j = Math.min(n, jobs.get(i).deadline) - 1; j >= 0; j--) {

                if (!times[j]) {

                    results[j] = i;
                    times[j] = true;
                    break;

                }

            }

        }

        for (int i = 0; i < n; i++) {
            if (times[i]) {
                System.out.println(jobs.get(results[i]));
            }
        }

    }

    public static void main(String[] args) {

        List<Job> jobs = new ArrayList<>();
        jobs.add(new Job("A", 6, 100));
        jobs.add(new Job("B", 2, 100));
        jobs.add(new Job("C", 1, 5));
        jobs.add(new Job("D", 1, 25));
        jobs.add(new Job("E", 3, 90));
        jobs.add(new Job("F", 2, 70));

        printBestOrder(jobs);

    }

}

File: Algorithms/Greedy/Jobs.js
"use strict";

class Job {

    constructor(n, d, p) {

        this.name = n;
        this.deadline = d;
        this.profit = p;

    }

    toString() {

        return this.name;

    }

}

let printBestOrder = (jobs) => {

    jobs.sort((a, b) => b.profit - a.profit);

    let n = jobs.length;

    let results = new Array(n),
        times = new Array(n);

    for (let i in jobs) {

        for (let j = Math.min(n, jobs[i].deadline) - 1; j >= 0; j--) {

            if (!times[j]) {

                results[j] = i;
                times[j] = true;

                break;

            }

        }

    }

    for (let i in jobs) {

        if (times[i]) {

            console.log(jobs[results[i]].toString());

        }

    }

}

let jobs = [
    new Job("A", 6, 100),
    new Job("B", 2, 100),
    new Job("C", 1, 5),
    new Job("D", 1, 25),
    new Job("E", 3, 90),
    new Job("F", 2, 70)
];

printBestOrder(jobs);
